test/puppetlabs/trapperkeeper/shutdown_test.clj

```diff
@@ -370,3 +372,64 @@
           app              (tk/boot-services-with-config [test-service] {})]
       (is (identical? app (check-for-errors! app))
           "app not returned for check-for-errors!"))))
+
+(deftest shutdown-during-restart-test
+  (testing "shutdown can't begin while restart or other lifecycle functions are in progress"
+    (let [first-stop-begun? (promise)
+          stop-should-proceed? (promise)
+          lifecycle-events (atom [])
+          svc (tk/service
+               [[:ShutdownService request-shutdown]]
+               (init [this context]
+                     (swap! lifecycle-events conj :init)
+                     context)
+               (start [this context]
+                      (swap! lifecycle-events conj :start)
+                      context)
+               (stop [this context]
+                     (swap! lifecycle-events conj :stop)
+                     (request-shutdown)
+                     (deliver first-stop-begun? true)
+                     @stop-should-proceed?
+                     context))
+          app (testutils/bootstrap-services-with-config
+               [svc]
+               {})
+          ;; this ensures that the 'main' shutdown logic will be runnable,
+          ;; and gives us a way to observe when it has completed.
+          app-main-thread (future (tk/run-app app))]
+
+      ;; now we trigger a restart, which will call 'stop' for the first time,
+      ;; which will request a shutdown but will block on the stop-should-proceed
+      ;; promise
+      (internal/restart-tk-apps [app])
+
+      ;; wait until we know that the shutdown has been requested
+      @first-stop-begun?
+
+      ;; at this point, the app's shutdown-reason-promise should be set, but
+      ;; the main thread should be blocked because our first 'stop' is blocked.
+      (is (realized? (:shutdown-reason-promise @(tk-app/app-context app))))
+      (is (not (realized? app-main-thread)))
+
+      ;; validate that the first three events are as expected (we're still blocked
+      ;; in the first 'stop').
+      (let [expected-lifecycle-events [:init :start :stop]]
+        (while (< (count @lifecycle-events) (count expected-lifecycle-events))
```

Comment:
Do we need the loop here?  Shouldn't `expected-lifecycle-events` have exactly the `:init`, `:start`, and `:stop` elements at the point the `first-stop-begun` promise is first delivered?  Could just do the lifecycle event assertion and the `not (realized? app-main-thread)` assertion one time instead of two here?